### PR TITLE
execute proposal

### LINKED ISSUE #58: implement proposal.execute function

### PR DESCRIPTION
**Pull request by jellegerbrandy**
shoudl fix #58

### CODE REVIEW

File: src/arc.ts [hunk 1]
```diff
@@ -214,29 +214,35 @@ export class Arc {
...
+          subscription ${query}
+        `
+      const zenObservable: ZenObservable<object[]> = this.apolloClient.subscribe<object[]>({ query: subscriptionQuery })
+      const subscriptionObservable = Observable.create((obs: Observer<any>) => {
+          const subscription = zenObservable.subscribe(obs)
+          return () => subscription.unsubscribe()
+        })
 
-    return queryObservable as Observable<any>
+      const sub = from(queryPromise)
```
Comment (tibetsprague) [inline src/arc.ts]: this code is confusing, would be nice to have more comments throughout

File: src/arc.ts [hunk 2]
```diff
@@ -259,6 +265,10 @@ export class Arc {
         contractClass = require('@daostack/arc/build/contracts/ContributionReward.json')
         contract = new this.web3.eth.Contract(contractClass.abi, addresses.base.ContributionReward, opts)
         return contract
+      case 'GEN':
```
Comment (tibetsprague) [inline src/arc.ts]: why duplicate this code as GEN and DAOToken?

File: src/dao.ts [hunk 3]
```diff
@@ -122,6 +123,20 @@ export class DAO implements IStateful<IDAOState> {
     return this.context.getBalance(this.address)
   }
 
+  public approveForStaking(amount: number) {
```
Comment (tibetsprague) [inline src/dao.ts]: doesn't really make sense for this to be in dao.ts since approving GEN for staking is not specific to a DAO

File: src/dao.ts [hunk 4]
```diff
@@ -122,6 +123,20 @@ export class DAO implements IStateful<IDAOState> {
...
+    return this.context.GENToken().approveForStaking(amount)
+  }
+  /*
+   * return the allownace on the GEN conract for spender is GenesisProtocol
```
Comment (tibetsprague) [inline src/dao.ts]: mispelling and confusing comment

File: src/dao.ts [hunk 5]
```diff
@@ -122,6 +123,20 @@ export class DAO implements IStateful<IDAOState> {
...
+   */
+  public allowance(owner: string): Observable < any > {
```
Comment (tibetsprague) [inline src/dao.ts]: also not really Dao specific

File: src/proposal.ts [hunk 6]
```diff
@@ -17,11 +18,12 @@ export enum ProposalOutcome {
...
 
 export enum ProposalStage {
-  Open,
-  Boosted,
-  QuietEndingPeriod,
+  ExpiredInQueue,
+  Executed,
   Queued,
-  Resolved
+  PreBoosted,
```
Comment (tibetsprague) [inline src/proposal.ts]: so there isn't another stage for Pending Boost, its still PreBoosted but just has enough gens to be boosted after pending time period passes?

File: src/proposal.ts [hunk 7]
```diff
@@ -166,52 +172,53 @@ export class Proposal implements IStateful<IProposalState> {
...
           preBoostedVotePeriodLimit
           thresholdConst
           limitExponentValue
           quietEndingPeriod
           proposingRepReward
           votersReputationLossRatio
           minimumDaoBounty
-          daoBountyConst
-          activationTime
+          externalToken
```
Comment (tibetsprague) [inline src/proposal.ts]: we did you start doing alphabetical listing here and then stop?

File: src/proposal.ts [hunk 8]
```diff
@@ -269,32 +259,47 @@ export class Proposal implements IStateful<IProposalState> {
...
     // (does not work with new contract versions anymore, though, it seems)
     // const contributionRewardInstance = this.dao.getContract('ContributionReward')
     // const result = await contributionRewardInstance.methods.parameters(this.dao.address).call()
     // const votingMachineAddress = (
     //   await contributionRewardInstance.methods.getSchemeParameters(daoAvatarAddress)).votingMachineAddress
-
     // the graph indexes it at contributionRewardProposal.votingMachine, but not on the proposal entity
-    // const votingMachine = this.dao.getContract('AbsoluteVote')
-    const votingMachine = this.context.getContract('GenesisProtocol')
+    return this.context.getContract('GenesisProtocol')
```
Comment (tibetsprague) [inline src/proposal.ts]: hmm, with this code we will only support the latest deployed GenesisProtocol, just good to be aware of this

File: src/proposal.ts [hunk 9]
```diff
@@ -269,32 +259,47 @@ export class Proposal implements IStateful<IProposalState> {
...
+  public votes(options: IVoteQueryOptions = {}): Observable<IVote[]> {
+    options.proposal = this.id
+    return Vote.search(this.context, options)
+  }
+
+  /**
+   * Vote for this proposal
+   * @param  outcome one of ProposalOutcome.Pass (0) or ProposalOutcome.FAIL (1)
+   * @param  amount the amount of reputation to vote with. Defaults to 0 - in that case,
+   *  all the sender's rep will be used
```
Comment (tibetsprague) [inline src/proposal.ts]: I thought all the sender's rep was always used? why would you ever vote with less than all rep?

File: src/proposal.ts [hunk 10]
```diff
@@ -340,16 +345,18 @@ export class Proposal implements IStateful<IProposalState> {
     )
   }
 
+  public stakingToken() {
+    // TODO: the token to stake with from the proposal, instead ot he generic one
```
Comment (tibetsprague) [inline src/proposal.ts]: there are a lot of TODOs in here, have you created issues for all of them to remember?

File: src/proposal.ts [hunk 11]
```diff
@@ -340,16 +345,18 @@ export class Proposal implements IStateful<IProposalState> {
...
 
   public stake(outcome: ProposalOutcome, amount: number ): Operation<Stake> {
-    // TODO: cf. vote() function: get the contract from the proposal
-    const votingMachine = this.context.getContract('GenesisProtocol')
-
-    const stakeMethod = votingMachine.methods.stake(
+    const stakeMethod = this.votingMachine().methods.stake(
       this.id,  // proposalId
       outcome, // a value between 0 to and the proposal number of choices.
       amount // amount the amount to stake with . if _amount == 0 it will use all staker reputation.
```
Comment (tibetsprague) [inline src/proposal.ts]: reputation has nothing to do with staking, and you always have to choose an amount to stake

File: src/proposal.ts [hunk 12]
```diff
@@ -406,10 +416,28 @@ export class Proposal implements IStateful<IProposalState> {
...
   }
 
-  public rewards(options: IRewardQueryOptions = {}): Observable < IRewardState[] > {
+  public rewards(options: IRewardQueryOptions = {}): Observable<IRewardState[]> {
     options.proposal = this.id
     return Reward.search(this.context, options)
   }
+
+  public claimRewards(account: Address): Operation<boolean> {
+    const transaction = this.votingMachine().methods.redeem(this.id, account)
```
Comment (tibetsprague) [inline src/proposal.ts]: im pretty sure we need to use the Redeemer contract to redeem all the rewards at once, not just the votingMachine redeem

File: src/arc.ts (hunk 2 above)
Comment (jellegerbrandy) [reply src/arc.ts]: it 's in anticipation of a change in the migration repository which deploys two separate contracts...

File: src/dao.ts (hunk 3 above)
Comment (jellegerbrandy) [reply src/dao.ts]: good point

File: src/proposal.ts (hunk 8 above)
Comment (jellegerbrandy) [reply src/proposal.ts]: yes, indeed

File: src/proposal.ts (hunk 9 above)
Comment (jellegerbrandy) [reply src/proposal.ts]: i'm not sure. The solidity code has this param, I just copied it over.

File: src/proposal.ts (hunk 12 above)
Comment (jellegerbrandy) [reply src/proposal.ts]: :+1:

File: src/arc.ts (hunk 1 above)
Comment (jellegerbrandy) [reply src/arc.ts]: ok

File: src/dao.ts (hunk 5 above)
Comment (jellegerbrandy) [reply src/dao.ts]: moved it to the arc namespace

File: src/proposal.ts (hunk 11 above)
Comment (jellegerbrandy) [reply src/proposal.ts]: true

File: src/proposal.ts (hunk 12 above)
Comment (jellegerbrandy) [reply src/proposal.ts]: created a seperate issue for that: https://github.com/daostack/client/issues/84

File: src/proposal.ts (hunk 10 above)
Comment (jellegerbrandy) [reply src/proposal.ts]: I've addressed some of them;added some issues (both here and in the subgraph) for them; the ones that are left are mostly commens that are more about the code than about fucntionality ("write this test"), and I've just left them there..

**Comment by jellegerbrandy** [comment]
ok I see it now, a very stupid bug indeed, sorry about that. (some client test should have caught this)

**Pull request merged**